### PR TITLE
[#11] [API] As a user, I can sign in

### LINKED ISSUE #11: [API] As a user, I can sign in
## Acceptance Criteria

- Create a new endpoint to sign in: `POST api/v1/auth/sign-in`

  - Request:
  ```
  {
    "client_id": "id,
    "client_secret": "secret",
    "email": "markg@gmail.com",
    "password": "xxxx",
  }
  ```
  - Response: Token 

- Token
```
{
  "type": "token",
  "attributes":  {
    "access_token": "access token",
    "token_type": "bearer",
    "expires_in": 7200,
    "refresh_token": "refresh token",
    "scope": "scopes"
  }
}
```

- Using the query from #26 to get the response object

### PR DESCRIPTION
**Pull request by markgravity**
- Close #11

## What happened 👀

- Add `sign-in` endpoint to the router
- Add `SignIn` to `auth` controller and write tests for it

## Insight 📝

n/a

## Proof Of Work 📹
CleanShot 2023-08-23 at 14 22 31

### CODE REVIEW

File: lib/api/v1/controllers/auth_test.go [hunk 1]
```diff
@@ -75,4 +75,88 @@ var _ = Describe("AuthController", func() {
...
+				controller := controllers.AuthController{}
+
+				controller.SignIn(ctx)
+
+				Expect(resp.Result()).To(test.MatchJSONSchema("response/token/success"))
+			})
+		})
+
+		Context("Given INVALID payload", func() {
+			It("returns a bad request status", func() {
```
Comment (malparty) [inline lib/api/v1/controllers/auth_test.go]:
```suggestion
			It("returns the bad request status", func() {
```

Comment (malparty) [reply lib/api/v1/controllers/auth_test.go]: (same for `It("returns status OK", func() ` above 👍 )

**Review by malparty — APPROVED** [review]
Approved ahead with just 2 comments (similar to the previous PR) 👍

File: lib/api/v1/controllers/auth.go [hunk 2]
```diff
@@ -30,3 +32,26 @@ func (AuthController) SignUp(ctx *gin.Context) {
 
 	jsonhelpers.RenderJSON(ctx, http.StatusOK, nil)
 }
+
+func (AuthController) SignIn(ctx *gin.Context) {
+	server := oauth.GetOAuthServer()
+
+	gt, tgr, err := server.ValidationTokenRequest(ctx.Request)
```
Comment (suho) [inline lib/api/v1/controllers/auth.go]: Naming for variables are not clear, please improve it.

File: lib/api/v1/serializers/token.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/go-oauth2/oauth2/v4"
+	"github.com/google/uuid"
+)
+
+type TokenSerializer struct {
+	Token     oauth2.TokenInfo
+	TokenType string
+}
+
+func (s *TokenSerializer) Data() (response *responsemodels.TokenResponse) {
```
Comment (suho) [inline lib/api/v1/serializers/token.go:17]: Please write tests for this

**Comment by markgravity** [comment]
@suho solved at [3b61d41](https://github.com/markgravity/golang-ic/pull/42/commits/3b61d41d9a86b4dc3aacc01fb910b94543aba135) 🙏

**Review by suho — APPROVED** [review]

**Pull request merged**